server/src/main/java/org/apache/iotdb/db/engine/upgrade/UpgradeTask.java

```diff
@@ -30,40 +33,82 @@
 
   private final TsFileResource upgradeResource;
   private static final Logger logger = LoggerFactory.getLogger(UpgradeTask.class);
+  private static final String DOT_SEPERATOR = ",";
 
   public UpgradeTask(TsFileResource upgradeResource) {
     this.upgradeResource = upgradeResource;
   }
 
+
+  private static boolean writeUpgradeLogFile(RandomAccessFile upgradeLog, String content,
+      long position) {
+    UpgradeUtils.getUpgradeLogLock().writeLock().lock();
+    try {
+      upgradeLog.seek(position);
+      upgradeLog.write(content.getBytes());
+      upgradeLog.write(System.getProperty("line.separator").getBytes());
+      return true;
+    } catch (IOException e) {
+      logger.error("write upgrade log file failed, the log file:{}", upgradeLog);
+      return false;
+    } finally {
+      UpgradeUtils.getUpgradeLogLock().writeLock().unlock();
+    }
+  }
+
+  private static long getUpgradeLogLength(RandomAccessFile upgradeLog) {
+    UpgradeUtils.getUpgradeLogLock().readLock().lock();
+    try {
+      return upgradeLog.length();
+    } catch (IOException e) {
+      logger.error("read upgrade log file failed, the log file:{}", upgradeLog);
+      return 0;
+    } finally {
+      UpgradeUtils.getUpgradeLogLock().readLock().unlock();
+    }
+  }
+
   @Override
   public void run() {
-    try {
+    try (RandomAccessFile upgradeLogFile = new RandomAccessFile(UpgradeUtils.getUpgradeLogPath(),
+        "rw")) {
```

Comment:
Maybe you can use an append-only log file instead of RAF and append several logs for one task instead of writing in the same position, which can reduce the lock usages and seeks.

Reply:
I adopt your advice to use an append-only log file